yarn-project/p2p/src/service/tx_messages.ts

```diff
@@ -0,0 +1,205 @@
+ * @param message - The binary message to be decoded.
+ * @returns - The array of transactions originally encoded into the message.
+ */
+export function decodeTransactionsMessage(message: Buffer) {
+  const lengthSize = 4;
+  let offset = 0;
+  const txs: Tx[] = [];
+  while (offset < message.length) {
+    const dataSize = message.readUInt32BE(offset);
+    const totalSizeOfMessage = lengthSize + dataSize;
+    txs.push(fromTxMessage(message.subarray(offset, offset + totalSizeOfMessage)));
+    offset += totalSizeOfMessage;
+  }
+  return txs;
+}
+
+/**
+ * Create a POOLED_TRANSACTION_HASHES message.
+ * @param hashes - The transaction hashes to be sent.
+ * @returns The encoded message.
+ */
+export function createTransactionHashesMessage(hashes: TxHash[]) {
+  const messageData = hashes.map(x => x.buffer);
+  return createMessage(Messages.POOLED_TRANSACTION_HASHES, Buffer.concat(messageData));
+}
+
+/**
+ * Decode a POOLED_TRANSACTION_HASHESs message ito the original transaction hash objects.
+ * @param message - The binary message to be decoded.
+ * @returns - The array of transaction hashes originally encoded into the message.
+ */
+export function decodeTransactionHashesMessage(message: Buffer) {
+  let offset = 0;
+  const txHashes: TxHash[] = [];
+  while (offset < message.length) {
+    const slice = message.subarray(offset, offset + TxHash.SIZE);
+    if (slice.length < TxHash.SIZE) {
+      throw new Error(`Invalid message size when processing transaction hashes message`);
+    }
+    txHashes.push(new TxHash(slice));
+    offset += TxHash.SIZE;
+  }
+  return txHashes;
+}
+
+/**
+ * Create a GET_TRANSACTIONS message from an array of transaction hashes.
+ * @param hashes - The hashes of the transactions to be requested.
+ * @returns The encoded message.
+ */
+export function createGetTransactionsRequestMessage(hashes: TxHash[]) {
+  const messageData = hashes.map(x => x.buffer);
+  return createMessage(Messages.GET_TRANSACTIONS, Buffer.concat(messageData));
+}
+
+/**
+ * Decode a GET_TRANSACTIONS message into the original transaction hash objects.
+ * @param message - The binary message to be decoded.
+ * @returns - The array of transaction hashes originally encoded into the message.
+ */
+export function decodeGetTransactionsRequestMessage(message: Buffer) {
+  // for the time being this payload is effectively the same as the POOLED_TRANSACTION_HASHES message
+  return decodeTransactionHashesMessage(message);
+}
+
+/**
+ * Decode the message type from a received message.
+ * @param message - The received message.
+ * @returns The decoded MessageType.
+ */
+export function decodeMessageType(message: Buffer) {
+  return message.readUInt32BE(0);
+}
+
+/**
+ * Return the encoded message (minus the header) from received message buffer.
+ * @param message - The complete received message.
+ * @returns The encoded message, without the header.
+ */
+export function getEncodedMessage(message: Buffer) {
+  return message.subarray(4);
+}
+
+/**
```

Comment:
I created these serialisation functions as we didn't have anything for `Tx` objects. Could possibly just become `toBuffer` and `fromBuffer` on `Tx`. Would prefer to use the existing serialisation but we would need to handle undefined properties.

Reply:
> but we would need to handle undefined properties.

We could probably introduce a `Maybe` or `Optional` wrapper that just writes down an initial byte to indicate if the value is set or not, similar to how the `Vector` class writes the size first. 

However, since we're moving to msgpack now, maybe it makes sense to do things the msgpack way instead? We may want to loop in @ludamad here.